src/storage/src/render/sources.rs

```diff
@@ -199,247 +198,228 @@ where
+                                                    Subtime::least_summary(),
+                                                ),
+                                                metrics: backpressure_metrics.clone(),
+                                            }),
+                                            backpressure_metrics,
+                                        )
+                                    } else {
+                                        (None, None, None)
+                                    }
                                 } else {
-                                    (Collection::new(empty(scope)), None, None, None)
+                                    (None, None, None)
                                 };
-                            let (upsert, health_update, upsert_token) = crate::render::upsert::upsert(
-                                &upsert_input.enter(scope),
-                                upsert_envelope.clone(),
-                                refine_antichain(&resume_upper),
-                                previous,
-                                previous_token,
-                                base_source_config.clone(),
-                                &storage_state.instance_context,
-                                &storage_state.storage_configuration,
-                                &storage_state.dataflow_parameters,
-                                backpressure_metrics,
-                            );
-
-                            // Even though we register the `persist_sink` token at a top-level,
-                            // which will stop any data from being committed, we also register
-                            // a token for the `upsert` operator which may be in the middle of
-                            // rehydration processing the `persist_source` input above.
-                            needed_tokens.push(upsert_token);
-
-                            use mz_timely_util::probe::ProbeNotify;
-                            let handle = mz_timely_util::probe::Handle::default();
-                            let upsert = upsert.inner.probe_notify_with(vec![handle.clone()]);
-                            let probe = mz_timely_util::probe::source(
-                                scope.clone(),
-                                format!("upsert_probe({id})"),
-                                handle,
+                            let (stream, tok) = persist_source::persist_source_core(
+                                scope,
+                                id,
+                                persist_clients,
+                                description.ingestion_metadata,
+                                Some(as_of),
+                                SnapshotMode::Include,
+                                Antichain::new(),
+                                None,
+                                flow_control,
+                                false.then_some(|| unreachable!()),
```

Comment:
why do we have this??

Reply:
It's to help type inference. Passing just `None` doesn't compile since the type of the parameter is `Option<F> where F: Fn(..)` and so you must type out the generic parameter which of course you can't because it's a closure. So this is a trick to create an anonymous type that is never used.

Reply:
oh yeah thats sooo annoying